HIPStream.cu

```diff
@@ -0,0 +1,215 @@
+  check_error();
+  hipMemcpy(d_b, b.data(), b.size()*sizeof(T), hipMemcpyHostToDevice);
+  check_error();
+  hipMemcpy(d_c, c.data(), c.size()*sizeof(T), hipMemcpyHostToDevice);
+  check_error();
+}
+
+template <class T>
+void HIPStream<T>::read_arrays(std::vector<T>& a, std::vector<T>& b, std::vector<T>& c)
+{
+  // Copy device memory to host
+  hipMemcpy(a.data(), d_a, a.size()*sizeof(T), hipMemcpyDeviceToHost);
+  check_error();
+  hipMemcpy(b.data(), d_b, b.size()*sizeof(T), hipMemcpyDeviceToHost);
+  check_error();
+  hipMemcpy(c.data(), d_c, c.size()*sizeof(T), hipMemcpyDeviceToHost);
+  check_error();
+}
+
+
+template <typename T>
+__global__ void copy_kernel(hipLaunchParm lp, const T * a, T * c)
+{
+  const int i = hipBlockDim_x * hipBlockIdx_x + hipThreadIdx_x;
+  c[i] = a[i];
+}
+
+template <class T>
+void HIPStream<T>::copy()
+{
+  hipLaunchKernel(HIP_KERNEL_NAME(copy_kernel), dim3(array_size/TBSIZE), dim3(TBSIZE), 0, 0, d_a, d_c);
+  check_error();
+  hipDeviceSynchronize();
+  check_error();
+}
+
+template <typename T>
+__global__ void mul_kernel(hipLaunchParm lp, T * b, const T * c)
+{
+  const T scalar = 0.3;
+  const int i = hipBlockDim_x * hipBlockIdx_x + hipThreadIdx_x;
+  b[i] = scalar * c[i];
+}
+
+template <class T>
+void HIPStream<T>::mul()
+{
+  hipLaunchKernel(HIP_KERNEL_NAME(mul_kernel), dim3(array_size/TBSIZE), dim3(TBSIZE), 0, 0, d_b, d_c);
+  check_error();
+  hipDeviceSynchronize();
+  check_error();
+}
+
+template <typename T>
+__global__ void add_kernel(hipLaunchParm lp, const T * a, const T * b, T * c)
+{
+  const int i = hipBlockDim_x * hipBlockIdx_x + hipThreadIdx_x;
+  c[i] = a[i] + b[i];
+}
+
+template <class T>
+void HIPStream<T>::add()
+{
+  hipLaunchKernel(HIP_KERNEL_NAME(add_kernel), dim3(array_size/TBSIZE), dim3(TBSIZE), 0, 0, d_a, d_b, d_c);
+  check_error();
+  hipDeviceSynchronize();
+  check_error();
+}
+
+template <typename T>
+__global__ void triad_kernel(hipLaunchParm lp, T * a, const T * b, const T * c)
+{
+  const T scalar = 0.3;
+  const int i = hipBlockDim_x * hipBlockIdx_x + hipThreadIdx_x;
+  a[i] = b[i] + scalar * c[i];
+}
+
+template <class T>
+void HIPStream<T>::triad()
+{
+  hipLaunchKernel(HIP_KERNEL_NAME(triad_kernel), dim3(array_size/TBSIZE), dim3(TBSIZE), 0, 0, d_a, d_b, d_c);
+  check_error();
+  hipDeviceSynchronize();
+  check_error();
+}
+
+
+void listDevices(void)
+{
+  // Get number of devices
+  int count;
+  hipGetDeviceCount(&count);
+  check_error();
+
+  // Print device names
+  if (count == 0)
+  {
+    std::cerr << "No devices found." << std::endl;
+  }
+  else
+  {
+    std::cout << std::endl;
+    std::cout << "Devices:" << std::endl;
+    for (int i = 0; i < count; i++)
+    {
+      std::cout << i << ": " << getDeviceName(i) << std::endl;
+    }
+    std::cout << std::endl;
+  }
+}
+
+
+std::string getDeviceName(const int device)
+{
+  hipDeviceProp_t props;
+  hipGetDeviceProperties(&props, device);
+  check_error();
+  return std::string(props.name);
```

Comment:
Is there any way for HIP to get the actual device name (e.g. Fury X Nano, FirePro S9150 etc), rather than just the architecture name?

This isn't essential for this patch (AMD's OpenCL drivers also have this issue), I'm just curious.